synapse/dendrite/parser.py

```diff
@@ -0,0 +1,127 @@
+
+    def __init__(self, **kwargs):
+        self.opts = kwargs
+        self._assertOpts()
+        self.svcbus = s_telepath.openurl(self.opts.get('svcbus'))
+        self.proxy = s_service.SvcProxy(self.svcbus)
+        self.coordinator = self.proxy.getNameProxy(self.opts.get('coordinator'))
+        self.jobs = self.proxy.getNameProxy(self.opts.get('jobs'))
+
+    def register(self, mimeType, queue):
+        '''
+        Register a MIME type to queue name with the Dendrite Coordinator.
+
+        As a part of the registration process, a listener is attached to the ``'dendrite:jobs:%s' % queue`` in order
+        to detect when jobs become available for processing.
+
+        Args:
+            mimeType (str):
+                The MIME type as is returned by filemagic with the MAGIC_MIME_TYPE flag set. For more
+                information, see: https://filemagic.readthedocs.io/en/latest/api.html.
+
+            queue (str):
+                The name of the queue for the Coordinator to put jobs on.
+        '''
+        logger.info('registering %s => %s, and svcbus event: %s', mimeType, queue, self._eventName(queue))
+        self.coordinator.register(mimeType, queue)
+        self.svcbus.on(self._eventName(queue), self._processJob)
+
+    def unregister(self, mimeType, queue):
+        '''
+        Unregister a MIME type to queue name mapping with the Dendrite Coordinator.
+
+        Removes the MIME type to queue name mapping that the Coordinator uses to enqueue jobs. This has a system wide
+        impact and individual Parsers should use care when calling this method, as there may be other parser instances
+        running that expect the registration to exist.
+
+        Args:
+            mimeType (str):
+                The MIME type as is returned by filemagic with the MAGIC_MIME_TYPE flag set. For more
+                information, see: https://filemagic.readthedocs.io/en/latest/api.html.
+
+            queue (str):
+                The name of the queue to unregister.
+        '''
+        logger.info('unregistering %s => %s, and svcbus event: %s', mimeType, queue, self._eventName(queue))
+        self.coordinator.unregister(mimeType, queue)
+        self.svcbus.off(self._eventName(queue), self._processJob)
+
+    def process(self, queue, job):
+        '''
+        Process a parse job.
+
+        This method is to be overridden by sub-classes. The typical implementation should involve actions such as
+        reading the bytes from the Axon service, parse the bytes, and store results in a Cortex. If an exception 
+        occurs during the processing and the exception is allowed to propagate out of the process method, the job
+        will be re-enqueued for subsequent processing.
+
+        Args:
+            queue (str):
+                Name of the queue that the job came from.
+
+            job (dict):
+                Dict containing properties of the job.
+        '''
```

Comment:
In order to allow implementors a function to over-ride, but still potentially have our own "in band" control ( and easily keep the top level doc string ) we've tended to call "internal" functions ( def _process(self, ... ) ) from within the "top level" API.